src/client/opamPinCommand.ml

```diff
@@ -693,3 +693,117 @@ let list st ~short =
+
+(* Must not be contained in a package name, version, nor url *)
+let scan_sep = '^'
+
+let scan ~normalise ~recurse ?subpath url =
+  let open OpamStd.Option.Op in
+  let pins_of_dir dir =
+    OpamPinned.files_in_source ~recurse ?subpath dir
+    |> OpamStd.List.filter_map (fun (nf, opamf, sb) ->
+        let opam = OpamFile.OPAM.safe_read opamf in
+        match (nf ++ OpamFile.OPAM.name_opt opam) with
+        | Some name ->
+          Some (name, (OpamFile.OPAM.version_opt opam), sb)
+        | None ->
+          OpamConsole.warning "Can not retrieve a package name from %s"
+            (OpamFilename.to_string (OpamFile.filename opamf));
+          None)
+  in
+  let pins, cleanup =
+    match OpamUrl.local_dir url with
+    | Some dir -> pins_of_dir dir, None
+    | None ->
+      let pin_cache_dir = OpamRepositoryPath.pin_cache url in
+      let cleanup = fun () ->
+        OpamFilename.rmdir @@ OpamRepositoryPath.pin_cache_dir ()
+      in
+      let basename =
+        match OpamStd.String.split (OpamUrl.basename url) '.' with
+        | [] ->
+          OpamConsole.error_and_exit `Bad_arguments
+            "Can not retrieve a path from '%s'"
+            (OpamUrl.to_string url)
+        | b::_ -> b
+      in
+      try
+        let open OpamProcess.Job.Op in
+        OpamProcess.Job.run @@
+        OpamRepository.pull_tree
+          ~cache_dir:(OpamRepositoryPath.download_cache
+                        OpamStateConfig.(!r.root_dir))
+          basename pin_cache_dir [] [url] @@| function
+        | Not_available (_,u) ->
+          OpamConsole.error_and_exit `Sync_error
+            "Could not retrieve %s" u
+        | Result _ | Up_to_date _ ->
+          pins_of_dir pin_cache_dir, Some cleanup
+      with e -> OpamStd.Exn.finalise e cleanup
+  in
+  let finalise = OpamStd.Option.default (fun () -> ()) cleanup in
+  OpamStd.Exn.finally finalise @@ fun () ->
+  if normalise then
+    OpamConsole.msg "%s"
+      (OpamStd.List.concat_map "\n"
+         (fun (name, version, sb) ->
+            Printf.sprintf "%s%s%c%s%s"
+              (OpamPackage.Name.to_string name)
+              (OpamStd.Option.to_string
+                 (fun v -> "." ^OpamPackage.Version.to_string v) version)
+              scan_sep
+              (OpamUrl.to_string url)
+              (OpamStd.Option.to_string (fun sb ->
+                   (String.make 1 scan_sep) ^ sb) sb))
+         pins)
+  else
+    ["# Name"; "# Version"; "# Url" (*; "# Subpath"*)] ::
+    List.map (fun (name, version, _sb) ->
+        [ OpamPackage.Name.to_string name;
+          (version >>| OpamPackage.Version.to_string) +! "-";
+          OpamUrl.to_string url;
+          (*sb +! "-"*) ]) pins
+    |> OpamStd.Format.align_table
+    |> OpamConsole.print_table stdout ~sep:"  "
+
+let looks_like_normalised args =
+    List.for_all (fun s -> OpamStd.String.contains_char s scan_sep) args
+
+let parse_pins pins =
+  let separator = Re.char scan_sep in
+  let re =
+    Re.(compile @@ seq [
```

Comment:
Better to use `Re.whole_string` than `bol`/`eol` (should'nt it be `bos`/`eos`, btw, since we just match one element ?)